### PR TITLE
Code refactor

### PR DESCRIPTION
**Pull request by omer5yerushalmi**

### CODE REVIEW

File: api/v1alpha1/nfspvc_types.go [hunk 1]
```diff
@@ -29,25 +29,18 @@ type NfsPvcSpec struct {
 	// Capacity is for the size of the nfs
```
Comment (mzeevi) [inline api/v1alpha1/nfspvc_types.go:29]: The comments here are a bit strange. It should read something like:

```
// Capacity represents the size of the PVC, it should match the size of the NFS
```

File: api/v1alpha1/nfspvc_types.go [hunk 2]
```diff
@@ -29,25 +29,18 @@ type NfsPvcSpec struct {
...
 	Capacity corev1.ResourceList `json:"capacity,omitempty" protobuf:"bytes,1,rep,name=capacity,casttype=ResourceList,castkey=ResourceName"`
 	// Path is the path of the nfs volume
```
Comment (mzeevi) [inline api/v1alpha1/nfspvc_types.go:31]: Change the comments here as well and in other places in the API

File: api/v1alpha1/nfspvc_types.go [hunk 3]
```diff
@@ -29,25 +29,18 @@ type NfsPvcSpec struct {
...
-	Path string `json:"path"`
+	Path string `json:"path" protobuf:"bytes,2,opt,name=path"`
```
Comment (mzeevi) [inline api/v1alpha1/nfspvc_types.go:32]: Do we need the protobuf here? What does it even do?

File: api/v1alpha1/nfspvc_types.go [hunk 4]
```diff
@@ -29,25 +29,18 @@ type NfsPvcSpec struct {
...
-	Server string `json:"server"`
+	Server string `json:"server" protobuf:"bytes,1,opt,name=server"`
 }
 
 // NfsPvcStatus defines the observed state of NfsPvc
 type NfsPvcStatus struct {
 	// PvcStatus is the current status of the NfsPvc object
-	PvcStatus PvcCreationStatus `json:"pvcStatus"`
+	PvcPhase string `json:"pvcPhase,omitempty" protobuf:"bytes,3,opt,name=pvcPhase"`
+	PvPhase  string `json:"pvPhase,omitempty" protobuf:"bytes,3,opt,name=pvPhase"`
```
Comment (mzeevi) [inline api/v1alpha1/nfspvc_types.go:41]: Fix the comments in the API here

File: cmd/main.go [hunk 5]
```diff
@@ -89,6 +90,15 @@ func main() {
 		os.Exit(1)
 	}
 
+	if !utils.IsEnvironmentVariablesExist() {
+		setupLog.Error(err, "failed to get configuration environment variable")
+		os.Exit(1)
+	}
```
Comment (mzeevi) [inline cmd/main.go:96]: Change function name to `doesEnvironmentVariableExist`

File: cmd/main.go [hunk 6]
```diff
@@ -89,6 +90,15 @@ func main() {
...
+	if !utils.IsReclaimPolicyValid(os.Getenv("RECLAIM_POLICY")) {
```
Comment (sahar2339) [inline cmd/main.go:97]: Maybe put all the checks for environment variables into one function, and return false/true one time with the corresponding error message(like the flow of validating webhook for example)

File: internal/controller/nfspvc_controller.go [hunk 7]
```diff
@@ -78,339 +75,68 @@ func (r *NfsPvcReconciler) SetupWithManager(mgr ctrl.Manager) error {
...
+	err, deleted := finalizer_utils.HandleResourceDeletion(ctx, nfspvc, logger, r.Client)
 	if err != nil {
-		nfspvc.Status.PvcStatus = danaiov1alpha1.PvcStatusError
-		updateErr := r.Status().Update(ctx, nfspvc)
-		if updateErr != nil {
-			logger.Error(updateErr, "Failed to update NfsPvc status to error.")
+		if failedToCleanUpErr, ok := err.(*finalizer_utils.FailedCleanUpError); ok {
+			// This means the error is of type *FailedCleanUpError
+			logger.Info(fmt.Sprintf("failed to handle NfsPvc deletion: %s, so trying again in a few seconds", failedToCleanUpErr.Message))
+			return ctrl.Result{RequeueAfter: time.Second * 4}, nil
```
Comment (sahar2339) [inline internal/controller/nfspvc_controller.go:95]: put 4 in const

File: internal/controller/utils/common.go [hunk 8]
```diff
@@ -0,0 +1,29 @@
...
+)
+
+var AllowedReclaimPolicies = []corev1.PersistentVolumeReclaimPolicy{
+	corev1.PersistentVolumeReclaimRecycle,
+	corev1.PersistentVolumeReclaimDelete,
+	corev1.PersistentVolumeReclaimRetain,
+}
+
+func IsEnvironmentVariablesExist() bool {
+	storageClass := os.Getenv("STORAGE_CLASS")
```
Comment (sahar2339) [inline internal/controller/utils/common.go:18]: maybe put the environment variables in const

File: internal/controller/utils/finalizer/finalizer.go [hunk 9]
```diff
@@ -0,0 +1,120 @@
...
+	"k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/types"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+)
+
+const FinalizerDeletionNfsPVc = "nfspvc.dana.io/nfspvc-protection"
+
+type FailedCleanUpError struct {
+	Message string
```
Comment (sahar2339) [inline internal/controller/utils/finalizer/finalizer.go:19]: you are great

File: internal/controller/utils/finalizer/finalizer.go [hunk 10]
```diff
@@ -0,0 +1,120 @@
...
+}
+
+// HandleResourceDeletion ensures the deletion of the nfspvc
+func HandleResourceDeletion(ctx context.Context, nfspvc danaiov1alpha1.NfsPvc, log logr.Logger, k8sClient client.Client) (error, bool) {
+	if nfspvc.ObjectMeta.DeletionTimestamp != nil {
+		if controllerutil.ContainsFinalizer(&nfspvc, FinalizerDeletionNfsPVc) {
+			//check if the pv and the pvc are deleted
+			pvcDeleted := false
+			pvc := corev1.PersistentVolumeClaim{}
+			if err := k8sClient.Get(ctx, types.NamespacedName{Namespace: nfspvc.Namespace, Name: nfspvc.Name}, &pvc); err != nil {
```
Comment (sahar2339) [inline internal/controller/utils/finalizer/finalizer.go:34]: Maybe a util function to check if resource exists?

File: internal/controller/utils/status/nfspvc.go [hunk 11]
```diff
@@ -0,0 +1,88 @@
...
+
+	pvcPhase := getPvcStatus(ctx, nfspvc, k8sClient)
+	pvPhase := getPvStatus(ctx, nfspvc, k8sClient)
+
+	if pvcPhase != nfspvc.Status.PvcPhase || pvPhase != nfspvc.Status.PvPhase {
+		nfspvcObject.Status.PvcPhase = pvcPhase
+		nfspvcObject.Status.PvPhase = pvPhase
+
+		// Use retry on conflict to update the nfspvc status
+		err := retry.RetryOnConflict(retry.DefaultRetry, func() error {
```
Comment (sahar2339) [inline internal/controller/utils/status/nfspvc.go:40]: you are great

File: internal/controller/utils/sync/nfspvc.go [hunk 12]
```diff
@@ -0,0 +1,181 @@
...
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+const (
+	NfsPvcDanaLabel         = "nfspvc.dana.io/nfspvc-owner"
+	PvcBindStatusAnnotation = "pv.kubernetes.io/bind-completed"
+)
+
+var StorageClass = os.Getenv("STORAGE_CLASS")
+var ReclaimPolicy = os.Getenv("RECLAIM_POLICY")
```
Comment (sahar2339) [inline internal/controller/utils/sync/nfspvc.go:26]: const

File: internal/controller/nfspvc_controller.go [hunk 13]
```diff
@@ -78,339 +75,68 @@ func (r *NfsPvcReconciler) SetupWithManager(mgr ctrl.Manager) error {
...
-// This function responsible to handle errors in the reconcile in update the NfsPvc accordingly
-func (r *NfsPvcReconciler) handleReconcileError(ctx context.Context, nfspvc *danaiov1alpha1.NfsPvc, err error) {
-	logger := r.Log
+	err, deleted := finalizer_utils.HandleResourceDeletion(ctx, nfspvc, logger, r.Client)
 	if err != nil {
-		nfspvc.Status.PvcStatus = danaiov1alpha1.PvcStatusError
-		updateErr := r.Status().Update(ctx, nfspvc)
-		if updateErr != nil {
-			logger.Error(updateErr, "Failed to update NfsPvc status to error.")
+		if failedToCleanUpErr, ok := err.(*finalizer_utils.FailedCleanUpError); ok {
```
Comment (sahar2339) [inline internal/controller/nfspvc_controller.go:92]: maybe a function isError

**Review by sahar2339 — APPROVED** [review]

**Pull request merged**